arangod/VocBase/Methods/UpgradeTasks.cpp

```diff
@@ -155,38 +120,372 @@ bool UpgradeTasks::upgradeGeoIndexes(TRI_vocbase_t& vocbase,
         if (res.fail()) {
           LOG_TOPIC("5550a", ERR, Logger::STARTUP)
               << "Error upgrading geo indexes " << res.errorMessage();
-          return false;
+          throw res;
         }
       }
     }
   }
+}
 
-  return true;
+Result createSystemCollections(TRI_vocbase_t& vocbase,
+                               std::vector<std::shared_ptr<LogicalCollection>>& createdCollections) {
+  typedef std::function<void(std::shared_ptr<LogicalCollection> const&)> FuncCallback;
+  FuncCallback const noop = [](std::shared_ptr<LogicalCollection> const&) -> void {};
+
+  std::vector<CollectionCreationInfo> systemCollectionsToCreate;
+  // the order of systemCollections is important. If we're in _system db, the
+  // UsersCollection needs to be first, otherwise, the GraphsCollection must be first.
+  std::vector<std::string> systemCollections;
+
+  Result res;
+  std::shared_ptr<LogicalCollection> colToDistributeShardsLike;
+  if (vocbase.isSystem()) {
+    // we will use UsersCollection for distributeShardsLike
+    std::tie(res, colToDistributeShardsLike) =
+        methods::Collections::createSystem(vocbase, StaticStrings::UsersCollection, true);
+    if (!res.ok()) {
+      return res;
+    }
+    createdCollections.push_back(colToDistributeShardsLike);
+    systemCollections.push_back(StaticStrings::GraphsCollection);
+    if (StatisticsFeature::enabled()) {
+      systemCollections.push_back(StaticStrings::StatisticsCollection);
+      systemCollections.push_back(StaticStrings::Statistics15Collection);
+      systemCollections.push_back(StaticStrings::StatisticsRawCollection);
+    }
+  } else {
+    // we will use GraphsCollection for distributeShardsLike
+    std::tie(res, colToDistributeShardsLike) =
```

Comment:
See above comment for UsersCollection, if Grpahs exist we will have the same issue here.

Reply:
Same as above. Resolved without changes.